File: backend/src/main/kotlin/ru/itmo/db/coursework/api/mapping/MachineRecipeDtoMapper.kt
package ru.itmo.db.coursework.api.mapping

import org.springframework.stereotype.Service
import ru.itmo.db.coursework.api.dto.MachineRecipeDto
import ru.itmo.db.coursework.model.CircuitBoard
import ru.itmo.db.coursework.model.MachineRecipe
import ru.itmo.db.coursework.model.Tea

@Service
class MachineRecipeDtoMapper {
    fun fromDto(dto: MachineRecipeDto, circuitBoard: CircuitBoard, teaById: Map<Int, Tea>) = with(dto, {
        MachineRecipe(
                id = id,
                circuitBoard = circuitBoard,
                circuitBoardAmount = circuitBoardAmount,
                workTime = workTime,
                price = price,
                teas = teas.mapKeys { teaById[it.key]!! }
        )
    })

    fun toDto(model: MachineRecipe) = with(model, {
        MachineRecipeDto(
                id = id,
                circuitBoardId = circuitBoard.id,
                circuitBoardAmount = circuitBoardAmount,
                workTime = workTime,
                price = price,
                teas = teas.mapKeys { it.key.id }
        )
    })
}

File: frontend/src/api/mapping/MachineRecipeDtoMapper.ts
import MachineRecipeDto from "$src/api/dto/MachineRecipeDto";
import MachineRecipe from "$src/game/model/MachineRecipe";

export default class MachineRecipeDtoMapper {
  public static fromDto(dto: MachineRecipeDto) {
    return new MachineRecipe(dto.id, dto.circuit_board_id, dto.circuit_board_amount, dto.work_time, dto.price, dto.teas)
  }

  public static toDto(model: MachineRecipe) {
    return new MachineRecipeDto(model.id, model.circuitBoardId, model.circuitBoardAmount, model.workTime, model.price, model.teas)
  }
}